test/tests/functional/pbs_hook_execjob_end.py

```diff
@@ -0,0 +1,215 @@
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+from tests.functional import *
+
+
+class TestPbsExecjobEnd(TestFunctional):
+    """
+    This tests the feature in PBS that allows
+    execjob_end hook to execute such that
+    pbs_mom is not blocked upon execution.
+    """
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        self.attr = {'event': 'execjob_end', 'enabled': 'True', 'alarm': '50'}
+        self.hook_body = ("import pbs\n"
+                          "import time\n"
+                          "e = pbs.event()\n"
+                          "pbs.logjobmsg(e.job.id, \
+                                         'executed execjob_end hook')\n"
+                          "time.sleep(10)\n"
+                          "pbs.logjobmsg(e.job.id, \
+                                         'execjob_end hook ended')\n"
+                          "e.accept()\n")
+
+    def test_execjob_end_non_blocking(self):
+        """
+        Test to make sure that mom is unblocked and running
+        exechost_periodic hook while it's child process is executing
+        execjob_end hook.
+        """
+        hook_name = "execjob_end_logmsg"
+        self.server.create_import_hook(hook_name, self.attr, self.hook_body)
+        hook_name = "exechost_periodic_logmsg"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "pbs.logmsg(pbs.LOG_DEBUG, \
+                                 'executed exechost_periodic hook')\n"
+                     "e.accept()\n")
+        attr = {'event': 'exechost_periodic', 'freq': '3', 'enabled': 'True'}
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        self.server.create_import_hook(hook_name, attr, hook_body)
+        jid = self.server.submit(j)
+        (_, str1) = self.mom.log_match("Job;%s;executed execjob_end hook" %
+                                       jid, n=100, max_attempts=10, interval=2)
+        date_time1 = str1.split(";")[0]
+        epoch1 = int(time.mktime(time.strptime(
+            date_time1, '%m/%d/%Y %H:%M:%S')))
+        # following message should be logged while execjob_end hook is in sleep
+        (_, str1) = self.mom.log_match("executed exechost_periodic hook",
+                                       starttime=epoch1-1, endtime=epoch1+10,
+                                       n=100, max_attempts=10, interval=1)
+        date_time2 = str1.split(";")[0]
+        epoch2 = int(time.mktime(time.strptime(
+            date_time2, '%m/%d/%Y %H:%M:%S')))
+        (_, str1) = self.mom.log_match(
+            "Job;%s;execjob_end hook ended" %
+            jid, starttime=epoch2-1, n=100,
+            max_attempts=10, interval=2)
+        date_time3 = str1.split(";")[0]
+        self.logger.info(
```

Comment:
I think you should assert on the time difference between date_time1 and date_time2 to be 3 seconds or less (or maybe 5 seconds for slower machines but definitely less than 10)

Reply:
We are capturing date_time because we need to verify hook messages are logged in the order mentioned, to confirm mom is not blocked on execjob_end hook execution. The order is verified with the use of starttime and endtime parameters.

Reply:
In case of failure to find hook's log message in the given time period (starttime=epoch1-1 and endtime=epoch1+10) then test will fail with log message not found error. Only in case of success, we'll see the final log message. There's no need of asserting the time difference because starttime and endtime parameters in this test will ensure that hook's log message is matched in this interval.